Gradient Descent/gradient.cpp

```diff
@@ -0,0 +1,156 @@
+#include <iostream>
+#include <vector>
+#include <math.h> 
+#include <numeric>
+#include <cstdlib>
+
+using namespace std;
+
+const double epsilon = pow(10,-6); // Fix a small epsilon for finite difference scheme to calculate grad
+
+// This is the function to be minimised. At the moment it is f(x) = x^2
+double objfun(vector<double>& x){               // Takes a vector x...
+    double output_val = 0.0;
+    for(int i=0 ; i < x.size(); i++){
+        output_val += pow(x[i],2);}             //... and computes x_i^2 for each coordinate...
+    return output_val;                      //... and returns vector of calculation.
+}
+
+// This another function to be minimised. Uncomment out if
+// you want to use it. At the moment it is f(x) = x^2 + x^4 + 15
+//double objfun(vector<double>& x){               // Takes a vector x...
+//    double output_val = 0.0;
+//    for(int i=0 ; i < x.size(); i++){
+//        output_val += pow(x[i],2) + pow(x[i],4);}             //... and computes x_i^2 +x_i^4 + 15 for each coordinate...
+//    return output_val + 15;                      //... and returns vector of calculation.
```

Comment:
Again this returns a double, so is the sum of the function evaluated at each point. 

This isn't wrong, just need to be careful in our definitions of the functions as they map from R^N to R